libraries/Microsoft.Bot.Connector/Authentication/Government/MicrosoftGovernmentAppCredentials.cs

```diff
@@ -62,5 +67,72 @@ public override string OAuthEndpoint
         {
             get { return GovernmentAuthenticationConstants.ToChannelFromBotLoginUrl; }
         }
+
+        protected override Lazy<IAuthenticator> BuildIAuthenticator()
+        {
+            return new Lazy<IAuthenticator>(
+                () =>
+                {
+                    var credential = new ClientCredential(MicrosoftAppId, MicrosoftAppPassword);
+
+                    IAuthenticator Make(string authority, string scope)
+                        => new AdalAuthenticator(
+                            credential,
+                            new OAuthConfiguration() { Authority = authority, Scope = scope },
+                            this.CustomHttpClient,
+                            this.Logger);
+
+                    if (OAuthScope == GovernmentAuthenticationConstants.ToChannelFromBotOAuthScope)
+                    {
+                        var optionOld = Make(GovernmentAuthenticationConstants.ToChannelFromBotLoginUrl, GovernmentAuthenticationConstants.ToChannelFromBotOAuthScope);
+                        var optionNew = Make("https://login.microsoftonline.us/botframework.onmicrosoft.us", "https://api.botframework.onmicrosoft.us");
+
+                        return new Authenticators(optionOld, optionNew);
+                    }
+                    else
+                    {
+                        return Make(OAuthEndpoint, OAuthScope);
+                    }
+                },
+                LazyThreadSafetyMode.ExecutionAndPublication);
+        }
+
+        private sealed class Authenticators : IAuthenticator
+        {
+            private readonly IAuthenticator[] inners;
+
+            public Authenticators(params IAuthenticator[] inners)
+            {
+                this.inners = inners ?? throw new ArgumentNullException(nameof(inners));
+            }
+
+            async Task<AuthenticatorResult> IAuthenticator.GetTokenAsync(bool forceRefresh)
+            {
+                ExceptionDispatchInfo info = null;
+
+                for (int index = 0; index < this.inners.Length; ++index)
+                {
+                    var inner = this.inners[index];
+                    try
+                    {
+                        return await inner.GetTokenAsync(forceRefresh).ConfigureAwait(false);
```

Comment:
suggest remembering which inner works and using that for further calls. it's just extra work to always loop through

Reply:
Given this needs to be thread safe, I was opting for simplicity.